File: scripts/check-play-status.mjs
/**
 * Omniverse Labs - Node.js Google Play API Status Checker
 * Self-contained JWT bearer authentication with zero external dependencies.
 */

import fs from 'fs';
import path from 'path';
import crypto from 'crypto';
import https from 'https';

const PACKAGE_NAME = process.argv[2] || process.env.PLAY_CONSOLE_PACKAGE || 'com.omniverselabs.anotadordejuegos';
const KEY_PATH = process.env.PLAY_CONSOLE_JSON_KEY_PATH || path.resolve(process.cwd(), 'service_account.json');

function base64Url(data) {
  return Buffer.from(data).toString('base64url');
}

async function getAccessToken(keyData) {
  const now = Math.floor(Date.now() / 1000);
  const header = { alg: 'RS256', typ: 'JWT' };
  const claimSet = {
    iss: keyData.client_email,
    scope: 'https://www.googleapis.com/auth/androidpublisher',
    aud: 'https://oauth2.googleapis.com/token',
    exp: now + 3600,
    iat: now
  };

  const encodedHeader = base64Url(JSON.stringify(header));
  const encodedClaim = base64Url(JSON.stringify(claimSet));
  const signatureInput = `${encodedHeader}.${encodedClaim}`;

  const signer = crypto.createSign('RSA-SHA256');
  signer.update(signatureInput);
  const signature = signer.sign(keyData.private_key, 'base64url');

  const jwt = `${signatureInput}.${signature}`;

  const postData = new URLSearchParams({
    grant_type: 'urn:ietf:params:oauth:grant-type:jwt-bearer',
    assertion: jwt
  }).toString();

  return new Promise((resolve, reject) => {
    const req = https.request('https://oauth2.googleapis.com/token', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/x-www-form-urlencoded',
        'Content-Length': Buffer.byteLength(postData)
      }
    }, res => {
      let body = '';
      res.on('data', chunk => body += chunk);
      res.on('end', () => {
        if (res.statusCode >= 200 && res.statusCode < 300) {
          resolve(JSON.parse(body).access_token);
        } else {
          reject(new Error(`Failed to fetch token: ${body}`));
        }
      });
    });
    req.on('error', reject);
    req.write(postData);
    req.end();
  });
}

function apiRequest(url, token, method = 'GET', body = null) {
  return new Promise((resolve, reject) => {
    const headers = { 'Authorization': `Bearer ${token}` };
    let payload = null;
    if (body) {
      payload = JSON.stringify(body);
      headers['Content-Type'] = 'application/json';
      headers['Content-Length'] = Buffer.byteLength(payload);
    }
    const req = https.request(url, { method, headers }, res => {
      let responseBody = '';
      res.on('data', chunk => responseBody += chunk);
      res.on('end', () => {
        try {
          const json = responseBody ? JSON.parse(responseBody) : {};
          if (res.statusCode >= 200 && res.statusCode < 300) {
            resolve(json);
          } else {
            reject({ status: res.statusCode, data: json });
          }
        } catch (err) {
          reject({ status: res.statusCode, raw: responseBody });
        }
      });
    });
    req.on('error', reject);
    if (payload) req.write(payload);
    req.end();
  });
}

async function main() {
  console.log(`\n============================================================`);
  console.log(`   Omniverse Labs - Google Play API Status (Node.js)       `);
  console.log(`============================================================`);
  console.log(`📦 Package: ${PACKAGE_NAME}\n`);

  if (!fs.existsSync(KEY_PATH)) {
    console.error(`❌ Missing service account file: ${KEY_PATH}`);
    console.error(`   Provide it via PLAY_CONSOLE_JSON_KEY_PATH or save as service_account.json in the workspace root.\n`);
    process.exit(1);
  }

  const keyData = JSON.parse(fs.readFileSync(KEY_PATH, 'utf-8'));
  const token = await getAccessToken(keyData);

  // 1. Create Edit
  const edit = await apiRequest(`https://androidpublisher.googleapis.com/androidpublisher/v3/applications/${PACKAGE_NAME}/edits`, token, 'POST', {});
  const editId = edit.id;
  console.log(`✅ Connected to Google Play Developer API (Edit ID: ${editId})`);

  // 2. Query Tracks
  const tracksRes = await apiRequest(`https://androidpublisher.googleapis.com/androidpublisher/v3/applications/${PACKAGE_NAME}/edits/${editId}/tracks`, token);
  const tracks = tracksRes.tracks || [];

  console.log(`\n📊 Release Tracks:`);
  for (const t of tracks) {
    console.log(`\n  🔹 Track: [${t.track.toUpperCase()}]`);
    if (!t.releases || t.releases.length === 0) {
      console.log(`     - No active releases.`);
    } else {
      for (const r of t.releases) {
        console.log(`     • Version: ${r.name || 'N/A'} (Codes: ${(r.versionCodes || []).join(', ')})`);
        console.log(`       - Status: ${(r.status || 'N/A').toUpperCase()}`);
        if (r.status === 'inProgress' && r.userFraction) {
          console.log(`       - Rollout: ${(r.userFraction * 100).toFixed(1)}%`);
        }
      }
    }
  }

  // 3. Clean up
  await apiRequest(`https://androidpublisher.googleapis.com/androidpublisher/v3/applications/${PACKAGE_NAME}/edits/${editId}`, token, 'DELETE');
}

main().catch(err => {
  console.error('\n❌ Google Play API Error:', err.data?.error?.message || err);
  process.exit(1);
});

File: scripts/check-play-status.py
import os
import sys
import json
import argparse
import requests

if sys.platform == "win32":
    sys.stdout.reconfigure(encoding="utf-8")
from google.oauth2.credentials import Credentials
from google.auth.transport.requests import Request
from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

OAUTH_TOKEN_STORE = os.path.expanduser(r"~\.config\omniverse_labs\play_oauth_token.json")
SCOPES = ["https://www.googleapis.com/auth/androidpublisher"]


def get_oauth_service():
    if not os.path.exists(OAUTH_TOKEN_STORE):
        raise FileNotFoundError(f"OAuth token not found at: {OAUTH_TOKEN_STORE}")

    with open(OAUTH_TOKEN_STORE, "r") as f:
        data = json.load(f)

    creds = Credentials(
        token=data.get("access_token"),
        refresh_token=data.get("refresh_token"),
        token_uri="https://oauth2.googleapis.com/token",
        client_id=data.get("client_id"),
        client_secret=data.get("client_secret"),
        scopes=SCOPES,
    )

    if creds.expired or not creds.valid:
        creds.refresh(Request())
        data["access_token"] = creds.token
        with open(OAUTH_TOKEN_STORE, "w") as f:
            json.dump(data, f, indent=2)

    return build("androidpublisher", "v3", credentials=creds, cache_discovery=False)


def check_app_status(package_name: str):
    print("\n============================================================")
    print("   Omniverse Labs - Google Play Live Status Checker         ")
    print("============================================================")
    print(f"📦 Package: {package_name}\n")

    service = get_oauth_service()

    try:
        # 1. Create Edit
        edits = service.edits()
        edit_req = edits.insert(body={}, packageName=package_name)
        edit_res = edit_req.execute()
        edit_id = edit_res["id"]
        print(f"✅ Authenticated & Connected to Google Play Developer API (Session ID: {edit_id})")

        # 2. Query Tracks
        tracks_res = edits.tracks().list(packageName=package_name, editId=edit_id).execute()
        tracks = tracks_res.get("tracks", [])

        print(f"\n📊 Release Tracks Status:")
        for t in tracks:
            track_name = t.get("track", "Unknown").upper()
            releases = t.get("releases", [])
            print(f"\n  🔹 Track: [{track_name}]")
            if not releases:
                print("     - No releases currently in this track.")
            for r in releases:
                status = r.get("status", "N/A").upper()
                name = r.get("name", "N/A")
                version_codes = r.get("versionCodes", [])
                fraction = r.get("userFraction", 1.0) * 100
                print(f"     • Version Name: {name} (Version Code(s): {', '.join(map(str, version_codes))})")
                print(f"       - Status: {status}")
                if status == "INPROGRESS":
                    print(f"       - Rollout: {fraction:.1f}%")
                
                notes = r.get("releaseNotes", [])
                if notes:
                    for n in notes:
                        print(f"       - Release Notes [{n.get('language', 'und')}]: {n.get('text', '').strip()}")

        # 3. Query Store Listings
        try:
            listings_res = edits.listings().list(packageName=package_name, editId=edit_id).execute()
            listings = listings_res.get("listings", [])
            print(f"\n🌐 Store Listings Configured: {len(listings)}")
            for lst in listings:
                print(f"  • [{lst.get('language')}] '{lst.get('title')}'")
                if lst.get('shortDescription'):
                    print(f"    Short: {lst.get('shortDescription')}")
        except Exception:
            pass

        # Clean up
        try:
            edits.delete(packageName=package_name, editId=edit_id).execute()
        except Exception:
            pass

    except HttpError as e:
        error_content = json.loads(e.content.decode("utf-8"))
        error_msg = error_content.get("error", {}).get("message", str(e))
        error_code = e.resp.status
        print(f"\n❌ Google Play API Response ({error_code}): {error_msg}")
        if error_code == 404:
            print("   👉 Note: The package exists or needs an initial draft created in the Console.")
        elif error_code == 403:
            print(f"   👉 Note: Permission check failed: {error_msg}")


def main():
    pkg = sys.argv[1] if len(sys.argv) > 1 else "com.omniverselabs.anotadordejuegos"
    check_app_status(pkg)


if __name__ == "__main__":
    main()
